File: 프로그래머스/3/150365. 미로 탈출 명령어/미로 탈출 명령어.cpp
#include <string>
#include <vector>
#include <cstdlib>

using namespace std;

int dx[4] = {1, 0, 0, -1};
int dy[4] = {0, -1, 1, 0};
char dstr[4] = {'d', 'l', 'r', 'u'};
string answer = "";

int gn, gm, gr, gc, gk;

void dfs(int curX, int curY, int dist, string str) {
    if(answer != "") return; // 다른 곳에서 먼저 찾았을 때
    
    if(curX == gr && curY == gc && dist == gk) {
        answer = str;
        return;
    }
    
    if(dist >= gk) return; // 이동하는 거리를 넘게 이동할 때
    
    if(abs(gr - curX) + abs(gc - curY) > gk - dist) return; // 현재 위치에서 목적지로 남은 이동거리 내에서 갈 수 없을 때
    
    for(int dir = 0; dir < 4; dir++) {
        int nxtX = curX + dx[dir];
        int nxtY = curY + dy[dir];
        
        if(nxtX <= 0 || nxtX > gn || nxtY <= 0 || nxtY > gm) continue;
        
        dfs(nxtX, nxtY, dist + 1, str + dstr[dir]);
    }
}

string solution(int n, int m, int x, int y, int r, int c, int k) {
    int dist = abs(r - x) + abs(c - y);
    if(dist % 2 != k % 2 || dist > k) {
        return "impossible";
    }
    
    gn = n; gm = m; gr = r; gc = c; gk = k;
    
    dfs(x, y, 0, "");
    
    return answer;
}

File: 프로그래머스/3/150365. 미로 탈출 명령어/미로 탈출 명령어.py
from collections import deque

s = ["d", "l", "r", "u"]
dx = [1, 0, 0, -1]
dy = [0, -1, 1, 0]

def is_out(x, y, n, m):
    if x < 1 or x > n or y < 1 or y > m:
        return True
    return False

def solution(n, m, x, y, r, c, k):
    short_cut = abs(x - r) + abs(y - c)
    if short_cut > k or (short_cut % 2 != k % 2):
        return "impossible"
    
    answer = ""
    stack = deque()
    stack.append((x, y, 0, ""))
    
    while stack:
        curX, curY, curDist, curStr = stack.pop()
        
        if (curX, curY, curDist) == (r, c, k):
            answer = curStr
            break
            
        curLeftDist = abs(curX - r) + abs(curY - c)
        if curLeftDist > k - curDist:
            continue
            
        for dir in range(4):
            # 스택이니깐 역 사전 순으로
            nxtX = curX + dx[3 - dir]
            nxtY = curY + dy[3 - dir]
            
            if is_out(nxtX, nxtY, n, m):
                continue
            
            stack.append((nxtX, nxtY, curDist + 1, curStr + s[3 - dir]))
    
    return answer